ksql-engine/src/main/java/io/confluent/ksql/function/UdfTemplate.java

```diff
@@ -0,0 +1,137 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@SuppressWarnings("WeakerAccess") // used from generated code
+public final class UdfTemplate {
+
+  private static final Logger LOG = LoggerFactory.getLogger(UdfTemplate.class);
+
+  private UdfTemplate() { /* private constructor for utility class */ }
+
+  public static String generateCode(final Method method, final String obj) {
+    final Class<?>[] params = method.getParameterTypes();
+
+    final CodeBlock.Builder code = CodeBlock.builder();
+    for (int idx = 0; idx < params.length; idx++) {
+      final Class<?> param = params[idx];
+      code.addStatement("$T arg$L = ($T) $T.coerce(args, $T.class, $L)",
+                        Primitives.wrap(param),
+                        idx,
+                        Primitives.wrap(param),
+                        UdfTemplate.class,
+                        param,
+                        idx);
+    }
+
+    final String args = IntStream.range(0, params.length)
+        .mapToObj(i -> "arg" + i)
+        .collect(Collectors.joining(", "));
+
+    code.addStatement("return (($T) $L).$L($L)",
+                      method.getDeclaringClass(), obj, method.getName(), args);
+
+    final String codeString = code.build().toString();
+    LOG.trace("Generated code:\n" + codeString);
+    return codeString;
+  }
+
+  @SuppressWarnings({"unchecked"}) // used by code generator
+  public static <T> T coerce(
+      final Object[] args,
+      final Class<? extends T> clazz,
+      final int index) {
+    final Object arg = args[index];
+    if (arg == null) {
+      if (clazz.isPrimitive()) {
+        throw new KsqlFunctionException(
+            String.format(
+                "Can't coerce argument at index %d from null to a primitive type", index));
+      }
+      return null;
+    }
+
+    if (Primitives.wrap(clazz).isAssignableFrom(arg.getClass())) {
+      return (T) arg;
+    } else if (arg instanceof String) {
+      try {
+        return fromString((String) arg, clazz);
+      } catch (Exception e) {
+        throw new KsqlFunctionException(
+            String.format("Couldn't coerce string argument '\"args[%d]\"' to type %s",
+                          index, clazz));
+      }
+    } else if (arg instanceof Number) {
+      try {
+        return (T) fromNumber((Number) arg, clazz);
+      } catch (Exception e) {
+        throw new KsqlFunctionException(
+            String.format("Couldn't coerce numeric argument '\"args[%d]:(%s) %s\"' to type %s",
+                          index, arg.getClass(), arg, clazz));
+      }
+    }
+
+    throw new KsqlFunctionException(
+        String.format("Impossible to coerce (%s) %s into %s", arg.getClass(), arg, clazz));
+  }
+
+  private static Number fromNumber(final Number arg, final Class<?> clazz) {
+    if (Integer.class.isAssignableFrom(Primitives.wrap(clazz))) {
+      return arg.intValue();
+    } else if (Long.class.isAssignableFrom(Primitives.wrap(clazz))) {
```

Comment:
nit: `else` is redundant.

Reply:
without `else` checkstyle actually fails with NPPath complexity greater than 10 :( i'll keep the rest with the same style for consistency